### PR TITLE
feat: Add Intelligent Network Intrusion Detection System using Machine Learning-#1813

### LINKED ISSUE #1813: Intelligent Network Intrusion Detection System using Machine Learning
## Project Description

This project focuses on building a machine learning-based intrusion detection system capable of identifying malicious network activities such as DoS attacks, brute-force attacks, botnet traffic, infiltration attempts, and unauthorized access patterns from network traffic data.

The project will involve:

* Exploratory Data Analysis (EDA)
* Data preprocessing and feature engineering
* Handling class imbalance using SMOTE
* Feature selection and correlation analysis
* Comparative evaluation of multiple ML algorithms

## Models to be Compared

* Random Forest Classifier
* XGBoost Classifier
* LightGBM Classifier
* Support Vector Machine (SVM)

## Evaluation Metrics

* Accuracy
* Precision
* Recall
* F1-score
* ROC-AUC Score
* Confusion Matrix

## Dataset

CICIDS2017 Dataset
Kaggle: https://www.kaggle.com/datasets/dhoogla/cicids2017

## Additional Features

* SHAP-based explainability analysis
* Attack traffic visualization dashboards
* Real-time packet classification simulation
* Comparative inference speed analysis

Please assign me this project, i have a lot experience in building ML or DL models from scratch to edge deployable

### PR DESCRIPTION
**Pull request by Rakshak05**
### Description
This PR resolves #1813 by implementing an end-to-end Machine Learning-based Intrusion Detection System (IDS) following the schema and categories of the CICIDS2017 dataset.

### Key Implementations
1. **Interactive Streamlit Dashboard (`app.py`)**: Includes:
   - Model Performance Dashboard comparing evaluation metrics and confusion matrices.
   - Local Explainability visualizer using SHAP values.
   - Real-time network traffic packet stream classifier simulation.
   - Comparative Inference Speed Benchmark (microseconds per packet).
2. **ML Pipeline Notebook (`intrusion_detection_notebook.ipynb`)**: Demonstrates EDA, feature correlation heatmaps, data scaling, SMOTE balance resampling, multi-model evaluation, and SHAP explanations.
3. **Training Automation Script (`train_models.py`)**: Automates scaling, SMOTE oversampling, training of 4 algorithms (Random Forest, XGBoost, LightGBM, SVM), metrics log, and serialization of trained models & SHAP artifacts.
4. **CICIDS2017 Traffic Simulator (`dataset_generator.py`)**: Programmatically synthesizes representative, imbalanced network flows modeled on actual attack classes (Benign, DoS, Brute Force, Botnet, Infiltration).
5. **Project Requirements and Docs**: Set up `requirements.txt`, project-specific `README.md`, and added the project row to the root `README.md`.

### Evaluation Results (Trained Models)
- **Random Forest**: 100% Accuracy | 48.33 us/packet inference latency
- **XGBoost**: 100% Accuracy | 4.04 us/packet inference latency (Fastest)
- **LightGBM**: 99.92% Accuracy | 9.17 us/packet inference latency
- **SVM**: 99.75% Accuracy | 149.22 us/packet inference latency

### How to Run
1. Navigate to the project directory:
   ```bash
   cd "Intelligent Network Intrusion Detection System using Machine Learning"
   ```
2. Install the requirements:
   ```bash
   pip install -r requirements.txt
   ```
3. Run model training:
   ```bash
   python train_models.py
   ```

4. Run the Streamlit Dashboard:
   ```bash
   streamlit run app.py
   ```

### CODE REVIEW

File: Intelligent Network Intrusion Detection System using Machine Learning/intrusion_detection_notebook.ipynb
Comment (Sibam-Paul) [inline Intelligent Network Intrusion Detection System using Machine Learning/intrusion_detection_notebook.ipynb:1]: Run the cells

**Review by Sibam-Paul — CHANGES_REQUESTED** [review]

**Review by Sibam-Paul — APPROVED** [review]
